### PR TITLE
Introduce gateway auth key feature

### PR DESCRIPTION
**Pull request by nicolaszhang**
Introduce gateway auth key feature

## Description
Introduce gateway auth key feature by adding two new cmdlets:
New-AzureRmDataFactoryGatewayAuthKey
Get-AzureRmDataFactoryGatewayAuthKey

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by nicolaszhang** [comment]
@shahabhijeet , could you please take a look at this?

File: src/ResourceManager/DataFactories/AzureRM.DataFactories.psd1 [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '3.1.0'
+ModuleVersion = '3.2.0'
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/AzureRM.DataFactories.psd1]: @nicolaszhang please revert the change made to the module version; we will update the version during the release process

File: src/ResourceManager/DataFactories/ChangeLog.md [hunk 2]
```diff
@@ -19,6 +19,10 @@
 -->
 ## Current Release
 
+## Version 3.2.0
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/ChangeLog.md]: @nicolaszhang please remove the `## Version 3.2.0` header and place the notes that you added under the `## Current Release` header

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+using System.Security.Permissions;
+
+namespace Microsoft.Azure.Commands.DataFactories
+{
+    [Cmdlet(VerbsCommon.Get, Constants.GatewayAuthKey, DefaultParameterSetName = ByFactoryName), OutputType(typeof(PSDataFactoryGatewayAuthKey))]
+    public class GetAzureDataFactoryGatewayAuthKeyCommand : DataFactoryBaseCmdlet
+    {
+        [Parameter(ParameterSetName = ByFactoryObject, Position = 0, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The data factory object.")]
+        public PSDataFactory DataFactory { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang this parameter should be renamed to `InputObject`; this follows the guidelines recommended by the PowerShell Core team. You can add an alias to `DataFactory` so that users can still use that parameter name for this parameter.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+using System.Globalization;
+using System.Management.Automation;
+using System.Security.Permissions;
+
+namespace Microsoft.Azure.Commands.DataFactories
+{
+    [Cmdlet(VerbsCommon.Get, Constants.GatewayAuthKey, DefaultParameterSetName = ByFactoryName), OutputType(typeof(PSDataFactoryGatewayAuthKey))]
+    public class GetAzureDataFactoryGatewayAuthKeyCommand : DataFactoryBaseCmdlet
+    {
+        [Parameter(ParameterSetName = ByFactoryObject, Position = 0, Mandatory = true, ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang this should be `ValueFromPipeline=true` rather than `ValueFromPipelineByPropertyName`

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            HelpMessage = "The data factory object.")]
+        public PSDataFactory DataFactory { get; set; }
+
+        [Parameter(ParameterSetName = ByFactoryName, Position = 1, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The data factory name.")]
+        [ValidateNotNullOrEmpty]
+        public string DataFactoryName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang this parameter should be renamed to `Name`; this follows the guidelines recommend by the PowerShell Core team. You can add an alias to `DataFactoryName` so that users can still use that parameter name for this parameter.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.DataFactories
+{
+    [Cmdlet(VerbsCommon.New, Constants.GatewayAuthKey, DefaultParameterSetName = ByFactoryName), OutputType(typeof(PSDataFactoryGatewayAuthKey))]
+    public class NewAzureDataFactoryGatewayAuthKeyCommand : DataFactoryBaseCmdlet
+    {
+        [Parameter(ParameterSetName = ByFactoryObject, Position = 0, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The data factory object.")]
+        public PSDataFactory DataFactory { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment about `InputObject`

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+using Microsoft.Azure.Commands.DataFactories.Properties;
+using System.Globalization;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.DataFactories
+{
+    [Cmdlet(VerbsCommon.New, Constants.GatewayAuthKey, DefaultParameterSetName = ByFactoryName), OutputType(typeof(PSDataFactoryGatewayAuthKey))]
+    public class NewAzureDataFactoryGatewayAuthKeyCommand : DataFactoryBaseCmdlet
+    {
+        [Parameter(ParameterSetName = ByFactoryObject, Position = 0, Mandatory = true, ValueFromPipelineByPropertyName = true,
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment about `ValueFromPipeline`

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            HelpMessage = "The data factory object.")]
+        public PSDataFactory DataFactory { get; set; }
+
+        [Parameter(ParameterSetName = ByFactoryName, Position = 1, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The data factory name.")]
+        [ValidateNotNullOrEmpty]
+        public string DataFactoryName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment about `Name`

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+        [Parameter(Position = 2, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The data factory gateway name.")]
+        [ValidateNotNullOrEmpty]
+        public string GatewayName { get; set; }
+
+        [Parameter(Position = 3, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The name of gateway auth key to be regenerated, either 'key1' or 'key2'.")]
+        [ValidateNotNullOrEmpty]
+        public string KeyName { get; set; }
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang if the `KeyName` is only limited to the values of `key1` or `key2`, then you should use the `ValidateSet` attribute for this parameter, which will limit the values users can provide for the parameter, as well as allow them to tab through the possible values

File: src/ResourceManager/DataFactories/Commands.DataFactories/help/AzureRM.DataFactories.md [hunk 10]
```diff
@@ -50,8 +53,11 @@ Encrypts sensitive data.
 ### [New-AzureRmDataFactoryGateway](New-AzureRmDataFactoryGateway.md)
 Creates a gateway for an Azure Data Factory.
 
+### [New-AzureRmDataFactoryGatewayAuthKey](New-AzureRmDataFactoryGatewayAuthKey.md)
+Creates auth key for an Azure Data Factory Gateway.
+
 ### [New-AzureRmDataFactoryGatewayKey](New-AzureRmDataFactoryGatewayKey.md)
-Creates a gateway key for an Azure Data Factory.
+Creates a gateway key for an Azure Data Factory. This cmdlet is deprecated, and you should use **New-AzureRmDataFactoryGatewayAuthKey** instead.
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/help/AzureRM.DataFactories.md]: @nicolaszhang if this cmdlet is being deprecated, can we add a warning to the first line of this cmdlet's execution letting users know that they should use `New-AzureRmDataFactoryGatewayAuthKey` instead?

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.DataFactories.Models;
+using Microsoft.Azure.Commands.DataFactories.Properties;
+using System.Globalization;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.DataFactories
+{
+    [Cmdlet(VerbsCommon.New, Constants.GatewayAuthKey, DefaultParameterSetName = ByFactoryName), OutputType(typeof(PSDataFactoryGatewayAuthKey))]
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang this cmdlet should implement [`SupportsShouldProcess`](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess)

File: src/ResourceManager/DataFactories/AzureRM.DataFactories.psd1 (hunk 1 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/AzureRM.DataFactories.psd1]: OK.

File: src/ResourceManager/DataFactories/ChangeLog.md (hunk 2 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/ChangeLog.md]: OK.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 4 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: But I just need DataFactoryName and ResourceName and don't need access the whole object. 

From doc, 
**ValueFromPipeline**
A Boolean value that indicates true if the parameter can take values from the incoming pipeline objects. This setting is used if the cmdlet must access the complete input object.

Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: And actually I'm following current code in data factories. For example, [NewAzureDataFactoryGatewayKeyCommand](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayKeyCommand.cs)

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 3 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: And actually I'm following current code in data factories. For example, [NewAzureDataFactoryGatewayKeyCommand](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayKeyCommand.cs)

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 5 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: I think if renaming it to Name, user may consider it's the name of "AzureDataFactoryGatewayAuthKey", but it's not it's DataFactoryName. And there are also GatewayName property. 

And I'm also referring to current data factory code.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 7 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: Same as above.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 6 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: Same as above.

File: src/ResourceManager/DataFactories/Commands.DataFactories/help/AzureRM.DataFactories.md (hunk 10 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/help/AzureRM.DataFactories.md]: Sure.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 8 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: Same as above. There are DataFactoryName, GatewayName, etc, we'd better distinguish them.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 11 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: And actually I'm following current code in data factories. For example, [NewAzureDataFactoryGatewayKeyCommand](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayKeyCommand.cs)

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 4 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang the scenario you mentioned is definitely valid, but using `ValueFromPipelineByPropertyName` here doesn't make too much sense unless a user would be piping an object with **a property whose name is DataFactory**. If you switch this to `ValueFromPipeline` then it allows a user to pipe the resulting `PSDataFactory` object from a cmdlet (such as `Get-AzureRmDataFactory`) to this cmdlet. It looks like the cmdlet you mentioned that previously used this pattern was incorrect.

An additional way of doing what you want is to create a separate parameter set that contains parameters `Name` (alias to `DataFactoryName`) and `ResourceGroupName`, and a user can provide those instead of the entire `PSDataFactory` object (you can also give these parameters the `ValueFromPipelineByPropertyName` attribute)

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 3 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang for new cmdlets, we are trying to ensure that they follow the guidelines and patterns established by the PowerShell Core team, even if that means they go against a design previously used for cmdlets. Even if this parameter doesn't share the same name as the one from the cmdlet you mentioned, it can still have an alias to `DataFactory`, so users will still be able to use that name if they don't want to use `InputObject`.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 11 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang a `New-*` cmdlet does not need to implement `SupportsShouldProcess` if it is creating an in-memory object. It doesn't appear to be the case here because this cmdlet is making a call to the client to regenerate the key. Please let me know if that is incorrect.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 7 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 6 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 5 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: 👍

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 8 above)
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: 👍

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 4 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: OK, thanks for the comments.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 3 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: OK, thanks for the comments.

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 9 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: OK.

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1715/

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            HelpMessage = "The data factory gateway name.")]
+        [ValidateNotNullOrEmpty]
+        public string GatewayName { get; set; }
+
+        [EnvironmentPermission(SecurityAction.Demand, Unrestricted = true)]
+        public override void ExecuteCmdlet()
+        {
+            if (ParameterSetName == ByFactoryObject)
+            {
+                if (InputObject == null)
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang you can use the `ValidateNotNull` attribute for the `InputObject` parameter instead of using this if-statement

```cs
[Parameter(ParameterSetName = ByFactoryObject, Position = 0, Mandatory = true, ValueFromPipeline = true,
    HelpMessage = "The data factory object.")]
[Alias("DataFactory")]
[ValidateNotNull]
public PSDataFactory InputObject { get; set; }
```

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        public override void ExecuteCmdlet()
+        {
+            if (!ShouldProcess("New-AzureRmDataFactoryGatewayAuthKey"))
+            {
+                return;
+            }
+
+            if (ParameterSetName == ByFactoryObject)
+            {
+                if (InputObject == null)
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang same comment about `ValidateNotNull`

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+        [Parameter(Position = 3, Mandatory = true, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The name of gateway auth key to be regenerated, either 'key1' or 'key2'.")]
+        [ValidateNotNullOrEmpty]
+        [ValidateSet("key1", "key2", IgnoreCase = true)]
+        public string KeyName { get; set; }
+        
+        public override void ExecuteCmdlet()
+        {
+            if (!ShouldProcess("New-AzureRmDataFactoryGatewayAuthKey"))
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: @nicolaszhang can we move this if-statement to the end of the `ExecuteCmdlet` method and have it wrap the last two lines? It would look something like

```cs
if (ShouldProcess(this.DataFactoryName, "New Data Factory Gateway Auth Key")
{
    PSDataFactoryGatewayAuthKey gatewayKey = DataFactoryClient.RegenerateGatewayAuthKey(ResourceGroupName, DataFactoryName, GatewayName, param);
    WriteObject(gatewayKey);
}
```

File: src/ResourceManager/DataFactories/Commands.DataFactories.Test/ScenarioTests/DataFactoryGatewayTests.cs [hunk 15]
```diff
@@ -37,6 +37,12 @@ public void TestCreateDataFactoryGateway()
             RunPowerShellTest("Test-DataFactoryGateway");
         }
 
+        [Fact]
```
Comment (cormacpayne) [inline src/ResourceManager/DataFactories/Commands.DataFactories.Test/ScenarioTests/DataFactoryGatewayTests.cs]: @nicolaszhang this test is failing on the on-demand run because the json file was not added to the test csproj file, so it is not being copied over to the output directory. If you follow the pattern that you do [for other json files](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/DataFactories/Commands.DataFactories.Test/Commands.DataFactories.Test.csproj#L291-L293) and add the few lines to the csproj, then the build should be fine.

Also, would you mind making this test a check-in test so it runs during the regular build and not just on-demand? The test below this one has the traits that you need to make it a check-in test

```cs
[Trait(Category.AcceptanceType, Category.CheckIn)]
```

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 12 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/GetAzureDataFactoryGatewayAuthKeyCommand.cs]: 👍

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 13 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: 👍

File: src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs (hunk 14 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories/Gateway/NewAzureDataFactoryGatewayAuthKeyCommand.cs]: sure

File: src/ResourceManager/DataFactories/Commands.DataFactories.Test/ScenarioTests/DataFactoryGatewayTests.cs (hunk 15 above)
Comment (nicolaszhang) [inline src/ResourceManager/DataFactories/Commands.DataFactories.Test/ScenarioTests/DataFactoryGatewayTests.cs]: Oh, my bad... updated now.

**Comment by cormacpayne** [comment]
On-demand: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1720/
